src/daft-core/src/array/ops/utf8.rs

```diff
@@ -62,6 +62,91 @@ where
+    index: usize,
+    name: &str,
+) -> DaftResult<Utf8Array> {
+    let arrow_result = iter
+        .map(|(val, re)| match (val, re) {
+            (Some(val), Some(re)) => {
+                // https://docs.rs/regex/latest/regex/struct.Regex.html#method.captures
+                // regex::find is faster than regex::captures but only returns the full match, not the capture groups.
+                // So, use regex::find if index == 0, otherwise use regex::captures.
+                if index == 0 {
+                    Ok(re?.find(val).map(|m| m.as_str()))
+                } else {
+                    Ok(re?
+                        .captures(val)
+                        .and_then(|captures| captures.get(index))
+                        .map(|m| m.as_str()))
+                }
+            }
+            _ => Ok(None),
+        })
+        .collect::<DaftResult<arrow2::array::Utf8Array<i64>>>();
+
+    Ok(Utf8Array::from((name, Box::new(arrow_result?))))
+}
+
+fn regex_extract_all_matches<'a>(
+    iter: impl Iterator<Item = (Option<&'a str>, Option<Result<regex::Regex, regex::Error>>)>,
+    index: usize,
+    len: usize,
+    name: &str,
+) -> DaftResult<ListArray> {
+    let mut matches = arrow2::array::MutableUtf8Array::new();
+    let mut offsets = arrow2::offset::Offsets::new();
+    let mut validity = arrow2::bitmap::MutableBitmap::with_capacity(len);
+
+    for (val, re) in iter {
+        let mut num_matches = 0i64;
+        match (val, re) {
+            (Some(val), Some(re)) => {
+                // https://docs.rs/regex/latest/regex/struct.Regex.html#method.captures_iter
+                // regex::find_iter is faster than regex::captures_iter but only returns the full match, not the capture groups.
+                // So, use regex::find_iter if index == 0, otherwise use regex::captures.
+                if index == 0 {
+                    for m in re?.find_iter(val) {
+                        matches.push(Some(m.as_str()));
+                        num_matches += 1;
+                    }
+                } else {
+                    for captures in re?.captures_iter(val) {
+                        if let Some(capture) = captures.get(index) {
+                            matches.push(Some(capture.as_str()));
+                            num_matches += 1;
+                        }
+                    }
+                }
+                validity.push(true);
+            }
+            (_, _) => {
+                validity.push(false);
+            }
+        }
+        offsets.try_push(num_matches)?;
+    }
+
+    let matches: arrow2::array::Utf8Array<i64> = matches.into();
+    let offsets: arrow2::offset::OffsetsBuffer<i64> = offsets.into();
+    let validity: Option<arrow2::bitmap::Bitmap> = match validity.unset_bits() {
+        0 => None,
+        _ => Some(validity.into()),
+    };
+    let flat_child = Series::try_from((
+        "matches",
+        Box::new(matches) as Box<dyn arrow2::array::Array>,
```

Comment:
Is there no `matches.into()` that you can use here?

Reply:
Ah there's a `to_boxed()` method